### PR TITLE
feat: Make easier for Invocable instance to register a route

### PR DESCRIPTION
**Pull request by eob**
This PR contains a slight refactoring of `Invocable`'s route registration code to make it easier for an `Invocable` instance to (1) register routes on itself, and (2) register routes against arbitrary function pointers.

## Current Situation

At present, route registration is done at the class-level, inside `__init_subclass__`, by associating HTTP routes with the **names** of functions to be called (rather than a pointers to that function).

This creates a number of frictions:

- It's very challenging for an Invocable instance (rather than classdef) to register a route
- It's very challenging for a helper library to contribute routes that should be mixed-in to an Invocable 

For example, the following things are near-impossible with our class-only registration system:

- Dynamically adding a new transport to (e.g. Slack, Telegram, Discord) to an Invocable
- Dynamically registering a callback into some Tool
- Dynamically registering a set of internal routes that aid in some async featuresset, like Ask My Book

## This refactor

This refactor implements a slight refactor to make it easier to register routes at the instance level. It does not change the manner of current execution.

1) It consolidates route registration into a set of helper methods (before this code was spread across various function bodies)
2) It consolidates more code into the PackageSpec and MethodSpec classes (following the design pattern of FastAPI)

As a result, the `__init__` method of an Invocable could now, from the instance perspective:

- Initialize a transport that registers the `/slack_webhook` route
- Initialize a tool that registers some specific callback
- Initialize a helper suite, such as the "Ask My.." mix-in, could register a set of HTTP paths that are used to sequence the import, chunking, and embedding of new documents

### CODE REVIEW

**Review by EniasCailliau — DISMISSED** [review]
If tests green light we can merge. Made a few tiny suggestions.

File: src/steamship/base/package_spec.py [hunk 1]
```diff
@@ -69,6 +72,18 @@ class MethodSpec(CamelModel):
...
     config: Optional[Dict] = None
 
+    # Bindings
+    # ------------------------
+    # We support two different kinds of bindings since most of our routes are bound by class decorators before
+    # the instance object is available. The `func_name_binding` stores a method name to be resolved against some
+    # invocable object at apply-time. The `func_binding` stores an actual callable
+
+    # A bound function name to call, provided an object
+    func_name_binding: Optional[str] = Field(None, exclude=True, repr=False)
```
[the remark this answers is not in the document]
Comment (eob) [reply src/steamship/base/package_spec.py:82]: I'm OK with this change; mostly neutral -- will go for it if you think it's clearner.

File: src/steamship/base/package_spec.py [hunk 2]
```diff
@@ -152,6 +198,9 @@ class PackageSpec(CamelModel):
     # The list of methods the package exposes remotely
     methods: Optional[List[MethodSpec]] = None
```
[the remark this answers is not in the document]
Comment (eob) [reply src/steamship/base/package_spec.py:199]: Fine w/ me -- will do

File: src/steamship/base/package_spec.py [hunk 3]
```diff
@@ -172,14 +221,48 @@ def import_parent_methods(self, parent: Optional["PackageSpec"] = None):
...
             return
         for method in parent.methods:
-            self.methods.append(method.clone())
+            self.add_method(method.clone())
 
     def add_method(self, new_method: MethodSpec):
         """Add a method to the MethodSpec, overwriting the existing if it exists."""
+        # Register the O(1) lookup
+        if not self.method_mappings:
+            self.method_mappings = {}
```
[the remark this answers is not in the document]
Comment (eob) [reply src/steamship/base/package_spec.py:230]: I'd tried this and I think it interacted strangely with Pydantic -- but I could have been mis-reading some other error. Will try again.

**Review by eob — COMMENTED** [review]
Thanks for the notes -- will take a pass and re-submit!

File: src/steamship/base/package_spec.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eob) [reply src/steamship/base/package_spec.py:230]: Going to sit on this change b/c the defaultdict behavior + Pydantic leaves me guessing about what's going to happen -- so at least for now want to trade clarity of execution behavior for slightly more verbose code.

**Review by dkolas — APPROVED** [review]
Though I see the appeal of this especially around transports, I'm wary of allowing dynamic method bindings.  It feels like there is some potential serious confusion / unintended consequences. Call it a yellow-flag level of concern; if we start using this, we need to make sure people understand that it will only work if it gets called from `_init_`.

Have one concrete suggestion below that may mitigate some possible errors.

File: src/steamship/base/package_spec.py [hunk 4]
```diff
@@ -69,6 +72,18 @@ class MethodSpec(CamelModel):
     #       But if Pydantic sees that, it attempts to force all values to be str, which is wrong.
     config: Optional[Dict] = None
 
+    # Bindings
+    # ------------------------
+    # We support two different kinds of bindings since most of our routes are bound by class decorators before
+    # the instance object is available. The `func_name_binding` stores a method name to be resolved against some
+    # invocable object at apply-time. The `func_binding` stores an actual callable
```
[the remark this answers is not in the document]
Comment (eob) [reply src/steamship/base/package_spec.py]: At the time of registration no instance yet exists since it's done upon subclass definition

**Pull request merged**